### ISSUE TITLE
Wallet corruption in v0.19.2

### ISSUE THREAD

**Issue by gracenoah** [Bitcoin-ABC/bitcoin-abc#317]
### Describe the issue
Since upgrading to v0.19.2, our testing cluster has been causing bitcoin cash's wallet to get corrupted every few days. We have periodic smoke tests that run against regnet and regularly provision addresses, perform transactions and force kill the node every time a new deployment happens.

### Can you reliably reproduce the issue?
Sort of. It happens every few days.

#### If so, please list the steps to reproduce below:
1. Add addresses to the wallet, force kill the node, perform transactions, mine blocks all the same time

### Expected behaviour
It sholudn't be possible to corrupt the wallet under any circumstance.

### Actual behaviour
The wallet being corrupted multiple times. Many backup files left over.
```
-rw------- 1 root root  34M Apr 15 07:30 wallet.dat
-rw------- 1 root root  25M Apr  8 17:25 wallet.dat.1554744510.bak
-rw------- 1 root root  31M Apr  8 17:28 wallet.dat.1554746665.bak
-rw------- 1 root root  35M Apr  8 18:06 wallet.dat.1554748674.bak
-rw------- 1 root root  25M Apr 11 17:59 wallet.dat.1555005673.bak
-rw------- 1 root root  32M Apr 11 18:01 wallet.dat.1555013224.bak
-rw------- 1 root root  38M Apr 11 20:12 wallet.dat.1555026682.bak
-rw------- 1 root root  41M Apr 12 00:21 wallet.dat.1555041993.bak
-rw------- 1 root root  41M Apr 12 04:49 wallet.dat.1555045277.bak
-rw------- 1 root root  43M Apr 12 05:45 wallet.dat.1555048882.bak
-rw------- 1 root root  22M Apr 14 20:50 wallet.dat.1555275147.bak
-rw------- 1 root root  27M Apr 14 20:52 wallet.dat.1555312859.bak
-rw------- 1 root root  32M Apr 15 07:21 wallet.dat.1555313152.bak
```

### What version of bitcoin-abc are you using?
v0.19.2 official binary

### Machine specs:
- OS: GCP kubernetes cluster with CoS
- CPU: 1
- RAM: 3GB
- Disk size: 10 GB
- Disk Type (HD/SDD): SSD

### Any extra information that might be useful in the debugging process.
Before being force killed:
![image](https://user-images.githubusercontent.com/36455668/56117870-b7bbeb80-5f71-11e9-9e86-9e1648c757ad.png)

After starting again:
![image](https://user-images.githubusercontent.com/36455668/56117543-f8673500-5f70-11e9-8b9e-85b624fc43ba.png)
![image](https://user-images.githubusercontent.com/36455668/56117460-c9e95a00-5f70-11e9-88e6-76f46434e8c1.png)

```
$ cat wallets/db.log | tail
BDB3018 unknown: unwritable page 1 remaining in the cache after error 30
BDB1586 temporary open: /var/tmp/BDB00027: Read-only file system
BDB3014 unable to create temporary backing file
BDB3018 unknown: unwritable page 0 remaining in the cache after error 30
BDB1586 temporary open: /var/tmp/BDB00027: Read-only file system
BDB3014 unable to create temporary backing file
BDB3018 unknown: unwritable page 0 remaining in the cache after error 30
BDB1586 temporary open: /var/tmp/BDB00027: Read-only file system
BDB3014 unable to create temporary backing file
BDB3018 unknown: unwritable page 0 remaining in the cache after error 30
```
```
$ cat wallets/db.log | grep -v temporary | grep -v 'unwritable' | tail -n 100
BDB1075 Page 2193: incorrect next_pgno 4069 found in leaf chain (should be 1047)
BDB1076 Page 1047: incorrect prev_pgno 1616 found in leaf chain (should be 2193)
BDB3017 unable to allocate space from the buffer cache
BDB1075 Page 4555: incorrect next_pgno 2612 found in leaf chain (should be 207)
BDB1076 Page 207: incorrect prev_pgno 3089 found in leaf chain (should be 4555)
BDB3017 unable to allocate space from the buffer cache
BDB1075 Page 5045: incorrect next_pgno 3350 found in leaf chain (should be 2708)
BDB1076 Page 2708: incorrect prev_pgno 4215 found in leaf chain (should be 5045)
BDB3017 unable to allocate space from the buffer cache
BDB1075 Page 88: incorrect next_pgno 3238 found in leaf chain (should be 12)
BDB1076 Page 12: incorrect prev_pgno 3721 found in leaf chain (should be 88)
BDB3017 unable to allocate space from the buffer cache
BDB1075 Page 4169: incorrect next_pgno 1596 found in leaf chain (should be 4507)
BDB1076 Page 4507: incorrect prev_pgno 2451 found in leaf chain (should be 4169)
BDB3017 unable to allocate space from the buffer cache
BDB1075 Page 275: incorrect next_pgno 4266 found in leaf chain (should be 1386)
BDB1076 Page 1386: incorrect prev_pgno 218 found in leaf chain (should be 275)
BDB3017 unable to allocate space from the buffer cache
BDB1075 Page 4612: incorrect next_pgno 2438 found in leaf chain (should be 1041)
BDB1076 Page 1041: incorrect prev_pgno 4915 found in leaf chain (should be 4612)
BDB3017 unable to allocate space from the buffer cache
BDB1075 Page 2549: incorrect next_pgno 5272 found in leaf chain (should be 322)
BDB1076 Page 322: incorrect prev_pgno 4074 found in leaf chain (should be 2549)
BDB3017 unable to allocate space from the buffer cache
BDB1075 Page 5284: incorrect next_pgno 1757 found in leaf chain (should be 314)
BDB1076 Page 314: incorrect prev_pgno 149 found in leaf chain (should be 5284)
BDB3017 unable to allocate space from the buffer cache
BDB1075 Page 25: incorrect next_pgno 196 found in leaf chain (should be 313)
BDB1076 Page 313: incorrect prev_pgno 4563 found in leaf chain (should be 25)
BDB3017 unable to allocate space from the buffer cache
BDB1075 Page 3251: incorrect next_pgno 5731 found in leaf chain (should be 688)
BDB1076 Page 688: incorrect prev_pgno 3063 found in leaf chain (should be 3251)
BDB3017 unable to allocate space from the buffer cache
BDB1075 Page 688: incorrect next_pgno 5705 found in leaf chain (should be 1243)
BDB1076 Page 1243: incorrect prev_pgno 5406 found in leaf chain (should be 688)
BDB3017 unable to allocate space from the buffer cache
BDB3017 unable to allocate space from the buffer cache
BDB1091 Page 3240: unterminated leaf chain
wallet.dat: BDB0090 DB_VERIFY_BAD: Database verification failed
BDB3017 unable to allocate space from the buffer cache
wallet.dat: Input/output error
BDB3017 unable to allocate space from the buffer cache
wallet.dat: Input/output error
BDB3017 unable to allocate space from the buffer cache
wallet.dat: Input/output error
BDB3017 unable to allocate space from the buffer cache
wallet.dat: Input/output error
BDB3017 unable to allocate space from the buffer cache
wallet.dat: Input/output error
BDB3017 unable to allocate space from the buffer cache
wallet.dat: Input/output error
BDB3017 unable to allocate space from the buffer cache
wallet.dat: Input/output error
wallet.dat.1555045277.bak: BDB0090 DB_VERIFY_BAD: Database verification failed
BDB3017 unable to allocate space from the buffer cache
wallet.dat: Input/output error
BDB3017 unable to allocate space from the buffer cache
BDB1075 Page 1329: incorrect next_pgno 3649 found in leaf chain (should be 2504)
BDB1076 Page 2504: incorrect prev_pgno 5131 found in leaf chain (should be 1329)
BDB3017 unable to allocate space from the buffer cache
BDB1075 Page 1577: incorrect next_pgno 3566 found in leaf chain (should be 1375)
BDB1076 Page 1375: incorrect prev_pgno 2593 found in leaf chain (should be 1577)
BDB3017 unable to allocate space from the buffer cache
BDB1075 Page 4566: incorrect next_pgno 2442 found in leaf chain (should be 16)
BDB1076 Page 16: incorrect prev_pgno 2584 found in leaf chain (should be 4566)
BDB3017 unable to allocate space from the buffer cache
BDB1075 Page 4246: incorrect next_pgno 1279 found in leaf chain (should be 45)
BDB1076 Page 45: incorrect prev_pgno 3372 found in leaf chain (should be 4246)
BDB3017 unable to allocate space from the buffer cache
BDB1075 Page 45: incorrect next_pgno 4717 found in leaf chain (should be 3738)
BDB1076 Page 3738: incorrect prev_pgno 1456 found in leaf chain (should be 45)
BDB3017 unable to allocate space from the buffer cache
BDB1075 Page 691: incorrect next_pgno 4548 found in leaf chain (should be 681)
BDB1076 Page 681: incorrect prev_pgno 3827 found in leaf chain (should be 691)
BDB3017 unable to allocate space from the buffer cache
BDB1075 Page 2724: incorrect next_pgno 1588 found in leaf chain (should be 1331)
BDB1076 Page 1331: incorrect prev_pgno 2468 found in leaf chain (should be 2724)
BDB3017 unable to allocate space from the buffer cache
BDB1075 Page 4249: incorrect next_pgno 882 found in leaf chain (should be 158)
BDB1076 Page 158: incorrect prev_pgno 3194 found in leaf chain (should be 4249)
BDB3017 unable to allocate space from the buffer cache
BDB1075 Page 5394: incorrect next_pgno 1756 found in leaf chain (should be 1001)
BDB1076 Page 1001: incorrect prev_pgno 4229 found in leaf chain (should be 5394)
BDB3017 unable to allocate space from the buffer cache
BDB1075 Page 5419: incorrect next_pgno 2015 found in leaf chain (should be 248)
BDB1076 Page 248: incorrect prev_pgno 2470 found in leaf chain (should be 5419)
BDB3017 unable to allocate space from the buffer cache
BDB1075 Page 3671: incorrect next_pgno 2263 found in leaf chain (should be 3210)
BDB1076 Page 3210: incorrect prev_pgno 738 found in leaf chain (should be 3671)
BDB3017 unable to allocate space from the buffer cache
BDB3017 unable to allocate space from the buffer cache
BDB1091 Page 1806: unterminated leaf chain
wallet.dat: BDB0090 DB_VERIFY_BAD: Database verification failed
BDB3017 unable to allocate space from the buffer cache
wallet.dat: Input/output error
BDB3017 unable to allocate space from the buffer cache
wallet.dat: Input/output error
wallet.dat.1555313152.bak: BDB0090 DB_VERIFY_BAD: Database verification failed
BDB3017 unable to allocate space from the buffer cache
wallet.dat: Input/output error
```

**Comment by gracenoah** [comment]
I recently realized that our system has been mining blocks every 10 seconds due to an unrelated bug, so that, combined with ungraceful shutdown during deployments might have been the reason we started seeing the wallet corruption very often now. Maybe the issue is not related to any particular version of bch or even related to bch at all. We've seen wallet corruption before on bitcoind as well.

I don't know if the tmp directory not being writable is related or not.
